src/System.Composition.Runtime/src/System/Composition/Hosting/Core/CompositionContract.cs

```diff
@@ -186,8 +186,11 @@ internal static bool ConstraintEqual(IDictionary<string, object> first, IDiction
                 if (!second.TryGetValue(firstItem.Key, out object secondValue))
                     return false;
 
-                if (firstItem.Value == null && secondValue != null ||
-                    secondValue == null && firstItem.Value != null)
+                if (firstItem.Value == null)
+                {
+                    return secondValue == null;
+                }
+                else if (secondValue == null)
```

Comment:
NRE was thrown below in line 205 as we call `firstItem.Value.Equals(secondValue)`